lib/src/dart.dart

```diff
@@ -90,3 +80,78 @@ Future snapshot(Uri executableUrl, String snapshotPath,
         failure.stderr.join('\n'));
   }
 }
+
+class AnalysisSessionManager {
+  final List<AnalysisContext> _contexts = [];
+
+  /// Parse the file with the given [path] into AST.
+  CompilationUnit parse(String path) {
+    var parseResult = _getSession(path).getParsedAstSync(path);
+    if (parseResult.errors.isNotEmpty) {
+      throw new AnalyzerErrorGroup(parseResult.errors);
+    }
+    return parseResult.unit;
+  }
+
+  /// Return import and export directives in the file with the given [path].
+  List<UriBasedDirective> parseImportsAndExports(String path) {
+    var unit = parse(path);
+    var uriDirectives = <UriBasedDirective>[];
+    for (var directive in unit.directives) {
+      if (directive is UriBasedDirective) {
+        uriDirectives.add(directive);
+      }
+    }
+    return uriDirectives;
+  }
+
+  AnalysisSession _getSession(String path) {
+    _throwIfNotAbsolutePath(path);
+
+    for (var context in _contexts) {
+      if (context.contextRoot.isAnalyzed(path)) {
+        return context.currentSession;
+      }
+    }
+
+    var dirPath = p.dirname(path);
```

Comment:
would it be sensible to search upwards for an apparent package root? Will the current setup involve creating multiple contexts if I first analyze `package/lib/some/deep/path/file.dart` and `package/lib/some/deep/different/file2.dart`? What will the performance costs be?

Reply:
That's good question, thank you.

We don't discover package roots by looking "up", only "down".
But Pub knows `Package`, and we can pre-create analysis contexts for them, so any request for a file will be handled by an existing analysis context.